File: go/binary_heap/heap_sort.go
package binary_heap

// HeapSort should sort the provided array in non-decreasing fashion using HeapSort algorithm
func HeapSort(arr []int) {
	HeapInit(arr)
	for i := len(arr) - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		HeapDown(arr[:i], 0)
	}

	for l, r := 0, len(arr)-1; l < r; l, r = l+1, r-1 {
		arr[l], arr[r] = arr[r], arr[l]
	}
}

File: java/src/binary_heap/HeapSort.java
package binary_heap;

public class HeapSort {
	// heapSort should sort the provided array in non-decreasing fashion using heapSort algorithm
	public static void heapSort(int[] arr) {
//		throw new UnsupportedOperationException("TODO");

		HeapInit.heapInit(arr);
		for (var n = arr.length-1; n > 0; n--) {
			var tmp = arr[0]; arr[0] = arr[n]; arr[n] = tmp;

			HeapDown.heapDown(arr, n, 0);
		}

		for (int l = 0, r = arr.length-1; l < r; l++, r--) {
			var tmp = arr[l]; arr[l] = arr[r]; arr[r] = tmp;
		}
	}
}
